### PR TITLE
feat: move libp2p to network worker thread

### LINKED ISSUE #5447: Move libp2p to worker thread
**Rationale**

Main goals:
- Move network crypto to thread
- Isolate network to protect main thread of DOS attacks
- Allow to throttle network processor better to keep main thread event loop lag under control 

Also, for nodes subscribed to many subnets recent benchmarks show that network handling takes decent chunk CPU time, lagging the main thread and delaying other pipelines like block processing.

- Part of https://github.com/ChainSafe/lodestar/issues/4856
- WIP branch all-in-one https://github.com/ChainSafe/lodestar/pull/5229

**Architecture**

Simple and lame chart for general idea. libp2p instance is rooted on the TCP transport's socket. It pulls many components that are best together in the same thread:
- peer manager
- gossip
- reqresp

On the other hand, the main thread is the only thread with access to the state cache, so the network processor must remain there. For now we'll place the network queues in the main thread, but those could be placed in the worker.

image

**Interfaces**

To do A/B testing we want to have a flag to switch the libp2p instance from being in a worker thread or the main thread. So we need
- A) Main thread "public" facing network API with variable backend. Implements `NetworkPublic` and calls `NetworkCore` internally (either B or C).
- B) Internal network class with logic de-duplicated from either backend. Implements `NetworkCore` with actual logic
- C) Main to worker event based interface to wire A to B through worker. Implements `NetworkCore` and calls `NetworkCore` internally (B).

```ts
interface NetworkPublic extends NetworkCorePublic {
  // Functions using multiple reqresp / gossip methods
  publishBeaconBlockMaybeBlobs(signedBlock: BlockInput): Promise<void>;
  beaconBlocksMaybeBlobsByRange(peerId: PeerId, request: phase0.BeaconBlocksByRangeRequest): Promise<BlockInput[]>;
  beaconBlocksMaybeBlobsByRoot(peerId: PeerId, request: phase0.BeaconBlocksByRootRequest): Promise<BlockInput[]>;
  
  // ReqResp caller helpers
  // NOTE: Should map to a single fn to prevent boilerplate
  status(peerId: PeerId, request: phase0.Status): Promise<phase0.Status>;
  goodbye(peerId: PeerId, request: phase0.Goodbye): Promise<void>;
  ping(peerId: PeerId): Promise<phase0.Ping>;
  metadata(peerId: PeerId): Promise<allForks.Metadata>;
  beaconBlocksByRange(peerId: PeerId, request: phase0.BeaconBlocksByRangeRequest): Promise<allForks.SignedBeaconBlock[]>;
  beaconBlocksByRoot(peerId: PeerId, request: phase0.BeaconBlocksByRootRequest): Promise<allForks.SignedBeaconBlock[]>;
  blobsSidecarsByRange(peerId: PeerId, request: deneb.BlobsSidecarsByRangeRequest): Promise<deneb.BlobsSidecar[]>;
  beaconBlockAndBlobsSidecarByRoot(peerId: PeerId, request: deneb.BeaconBlockAndBlobsSidecarByRootRequest): Promise<deneb.SignedBeaconBlockAndBlobsSidecar[]>;
  lightClientBootstrap(peerId: PeerId, request: Uint8Array): Promise<allForks.LightClientBootstrap>;
  lightClientOptimisticUpdate(peerId: PeerId): Promise<allForks.LightClientOptimisticUpdate>;
  lightClientFinalityUpdate(peerId: PeerId): Promise<allForks.LightClientFinalityUpdate>;
  lightClientUpdatesByRange(peerId: PeerId, request: altair.LightClientUpdatesByRange): Promise<allForks.LightClientUpdate[]>;

  // Gossip publish helpers
  // NOTE: This functions could be stand-alone since each have usually one single caller
  publishBeaconBlock(signedBlock: allForks.SignedBeaconBlock): Promise<PublishResult>
  publishSignedBeaconBlockAndBlobsSidecar(item: deneb.SignedBeaconBlockAndBlobsSidecar): Promise<PublishResult>
  publishBeaconAggregateAndProof(aggregateAndProof: phase0.SignedAggregateAndProof): Promise<PublishResult>
  publishBeaconAttestation(attestation: phase0.Attestation, subnet: number): Promise<PublishResult>
  publishVoluntaryExit(voluntaryExit: phase0.SignedVoluntaryExit): Promise<PublishResult>
  publishBlsToExecutionChange(blsToExecutionChange: capella.SignedBLSToExecutionChange): Promise<PublishResult>
  publishProposerSlashing(proposerSlashing: phase0.ProposerSlashing): Promise<PublishResult>
  publishAttesterSlashing(attesterSlashing: phase0.AttesterSlashing): Promise<PublishResult>
  publishSyncCommitteeSignature(signature: altair.SyncCommitteeMessage, subnet: number): Promise<PublishResult>
  publishContributionAndProof(contributionAndProof: altair.SignedContributionAndProof): Promise<PublishResult>
  publishLightClientFinalityUpdate(lightClientFinalityUpdate: allForks.LightClientFinalityUpdate): Promise<PublishResult>
  publishLightClientOptimisticUpdate(lightClientOptimisitcUpdate: allForks.LightClientOptimisticUpdate): Promise<PublishResult>
}
```

```ts
interface NetworkCore extends NetworkCorePublic {
  // Service
  metrics(): Promise<string>;
  close(): Promise<void>;

  sendReqRespRequest(method, peerId: PeerId, request): ??;
  publishGossip<K extends GossipType>(topic: GossipTopicMap[K], object: GossipTypeMap[K], opts?: PublishOpts): Promise<number>;
}
```

```ts
interface NetworkCorePublic {
  getEnr(): Promise<SignableENR | undefined>;
  getMetadata(): Promise<altair.Metadata>;
  getConnectedPeers(): PeerId[];
  getConnectedPeerCount(): number;

  prepareBeaconCommitteeSubnet(subscriptions: CommitteeSubscription[]): Promise<void>;
  prepareSyncCommitteeSubnets(subscriptions: CommitteeSubscription[]): Promise<void>;
  reStatusPeers(peers: PeerId[]): Promise<void>;
  reportPeer(peer: PeerId, action: PeerAction, actionName: string): Promise<void>;

  subscribeGossipCoreTopics(): Promise<void>;
  unsubscribeGossipCoreTopics(): Promise<void>;
  isSubscribedToGossipCoreTopics(): boolean;

  connectToPeer(peer: PeerId, multiaddr: Multiaddr[]): Promise<void>;
  disconnectPeer(peer: PeerId): Promise<void>;
  dumpPeers(): Promise<routes.lodestar.LodestarNodePeer[]>;
  dumpPeer(peerIdStr: string): Promise<routes.lodestar.LodestarNodePeer | undefined>;
  dumpPeerScoreStats(): Promise<PeerScoreStats>;
  dumpGossipPeerScoreStats(): Promise<PeerScoreStatsDump>;
  dumpGossipQueue(gossipType: GossipType): Promise<PendingGossipsubMessage[]>;
  dumpDiscv5KadValues(): Promise<string[]>;
}
```

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

- Allow network to perform time sensitive actions (gossip heartbeat, pings, etc) timely. DOS protection sandboxing resource spikes. See https://github.com/ChainSafe/lodestar/issues/4856

**Goals**

- Merge first minimum viable solution, going for simplicity first and optimization latter
- Allow libp2p to be run in main thread or worker by switching a runtime flag

**Description**

TBD

**TODO**

- [x] Wire metrics on libp2p worker
- [x] Connect NetworkEventBus through the worker boundary
- [x] Complete reqresp async iterables via events
- [x] Handle serialization/deserialization of req/resp across worker boundary
- [x] ~Move discv5 worker out of libp2p worker, and connect them via main thread~
- [x] Logger in worker with user provided settings
- [ ] Add metrics to track async iterable bridge Maps
- [x] Add tests for all possible events and methods crossing the worker boundary to programmatically ensure structures are clonable. Otherwise runtime error and Lodestar dead
- [x] Add more tests for both usage modes (in worker, outside worker)

Closes https://github.com/ChainSafe/lodestar/issues/5447

### CODE REVIEW

File: packages/beacon-node/src/network/core/mainThreadNetworkCore.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  const gossip = base.gossip;
+  for (const [methodName, method] of Object.entries(gossip)) {
+    if (method instanceof Function) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-call
+      (networkCore as any)[methodName] = method.bind(gossip);
+    }
+  }
+
+  for (const [methodName, method] of Object.entries(base)) {
+    if (method instanceof Function) {
```
Comment (dapplion) [inline packages/beacon-node/src/network/core/mainThreadNetworkCore.ts]: I think it's best to have code repetition that is obvious than a smart solution like this to save a few lines. Verbose but repetitive patterns are very easy to maintainers at little extra cost if based on common primitives like fn that calls function

File: packages/beacon-node/src/network/core/workerNetworkCore.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      bindAddr: opts.bindAddr,
+      metrics: opts.metrics,
+      peerStoreDir: opts.peerStoreDir,
+      genesisTime,
+      activeValidatorCount: modules.activeValidatorCount,
+    };
+
+    const worker = new Worker("./worker.js", {workerData} as ConstructorParameters<typeof Worker>[1]);
+
+    const workerApi = await spawn<NetworkWorkerApi>(worker, {
```
Comment (dapplion) [inline packages/beacon-node/src/network/core/workerNetworkCore.ts]: @wemeetagain I was wondering, why don't we run this promise on the root of the module? What's the need to delay worker instantiation?

File: packages/beacon-node/src/network/core/workerNetworkCore.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      // A Lodestar Node may do very expensive task at start blocking the event loop and causing
+      // the initialization to timeout. The number below is big enough to almost disable the timeout
+      timeout: 5 * 60 * 1000,
+    });
+
+    workerApi.pendingGossipsubMessage().subscribe((data) => events.emit(NetworkEvent.pendingGossipsubMessage, data));
```
Comment (dapplion) [inline packages/beacon-node/src/network/core/workerNetworkCore.ts]: Difference to discv5, this worker will have much more events. Having to do this observable pattern for each sounds unnecessary. We could emit all network events on a single observable, and then emit to their respective topics on the other side

File: packages/beacon-node/src/network/core/mainThreadNetworkCore.ts (hunk 1 above)
Comment (wemeetagain) [reply packages/beacon-node/src/network/core/mainThreadNetworkCore.ts]: Sounds good, I was going that route first, then tried to get clever :joy_cat:

Comment (wemeetagain) [reply packages/beacon-node/src/network/core/mainThreadNetworkCore.ts]: will do that tomorrow if you want to work on other stuff

File: packages/beacon-node/src/network/core/workerNetworkCore.ts (hunk 2 above)
Comment (wemeetagain) [reply packages/beacon-node/src/network/core/workerNetworkCore.ts]: need to wait for workerData, also makes testing easier imo to avoid module import side effects

**Comment by wemeetagain** [comment]
> todo
>  Move discv5 worker out of libp2p worker, and connect them via main thread

why should this happen?

**Comment by dapplion** [comment]
@wemeetagain  Can a worker spawn another worker?

**Comment by wemeetagain** [comment]
> @wemeetagain Can a worker spawn another worker?

Afaik yes

**Comment by dapplion** [comment]
@wemeetagain regarding ReqResp

- Split the ReqResp interface into two parts: `IReqRespBeaconNodePeerManager` only used by peer manager so never has to cross worker boundary. `IReqRespBeaconNodeBeacon` has to cross worker boundary. Shitty temp names, let's fix it latter.
- Responses on `IReqRespBeaconNodeBeaconBytes` methods are kept encoded to be decoded only when in the main thread. That happens at the very top `Network` class to prevent code duplication with current hierarchy.
- For handling async generators through worker boundary see `AsyncIterableWorker` which communicates all necessary signals

**Comment by nazarhussain** [comment]
@dapplion I reviewed the changes majorly focused to the `ReqResp` changes. I found those changes well enough to understand and use able. 

image

The points I can't understand is why this effort is been done with very particular focus on the `network`. I believe if for the planning phase, we should expand our focus to a higher level and answer few more questions. 

1. Discuss which component can and can't be splitted? 
2. Which of interfaces may be more easily vulnerable for DOS? 
3. Which of interfaces are more impactful for a DOS attack? 

We may definitely start with network, but then the focus for foreseeable future will be clear. e.g. I foresee the similar arguments for the Rest interface as well, but can't align how this refactor would enable us to do that next.

**Comment by wemeetagain** [comment]
The network is the source of all untrusted and adversarial input so it holds the vast amount of the risk.
The other factor we're considering is minimal disturbance to existing interfaces, especially interfaces with consumers that rely on fast synchronous access. Most data access across threads will become async, if there is not some caching across threads or use of exotic techniques ala SharedArrayBuffer. So this another big consideration / constraint to think about when deciding what to split out and how. Third consideration would be codebase simplicity / maintainability and having a solid grasp of the cost/benefit of splitting out a module and how that splitting out relates to the rest of the codebase.

So the network is a natural candidate to split out first, with medium cost and high benefit. We can offload a lot of normal work and worst-case work from the main thread, and the network somewhat decoupled from need for fast access to our chain.  All connection management (esp in DoS cases), network crypto (handshakes, connection encryption/decryption), simple req/resp (ping, status, metadata) can be handled entirely off-main-thread.

My preference for further splitting is that we are careful to avoid splitting 'just because', and have a clear sense of the benefits.
IMO the next module that I see having clear benefit to split is backfill sync.

**Comment by dapplion** [comment]
Rebased branch on current unstable with almost everything implemented and wired.

Pushed previous branch into backup branch to review if necessary https://github.com/ChainSafe/lodestar/compare/dapplion/network-thread-may1

CC @wemeetagain

**Comment by dapplion** [comment]
Big remaining item is the logger in the thread
- How to instantiate the logger with same settings?
- Ignore file logging completely or duplicate transports?

CC @wemeetagain

**Comment by twoeths** [comment]
Attached is the profile of network thread from lg1k node 
[0505_lg1k_network_thread.cpuprofile.zip](https://github.com/ChainSafe/lodestar/files/11403820/0505_lg1k_network_thread.cpuprofile.zip)

I was not able to do that by Chrome Dev Tool but with VsCode

File: packages/beacon-node/test/unit/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  });
+});
+
+// export function getStateTypeFromBytes(
+//   config: ChainForkConfig,
+//   bytes: Buffer | Uint8Array
+// ): allForks.AllForksSSZTypes["BeaconState"] {
+//   const slot = bytesToInt(bytes.subarray(SLOT_BYTES_POSITION_IN_STATE, SLOT_BYTES_POSITION_IN_STATE + SLOT_BYTE_COUNT));
+//   return config.getForkTypes(slot).BeaconState;
+// }
```
Comment (dapplion) [inline packages/beacon-node/test/unit/index.test.ts]: @tuyennhv is this committed by mistake?

Comment (twoeths) [reply packages/beacon-node/test/unit/index.test.ts]: it's my bad @dapplion , removed it

**Review by wemeetagain — DISMISSED** [review]

**Review by wemeetagain — APPROVED** [review]

**Comment by dapplion** [comment]
:tada:  thanks @wemeetagain :heart:

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.9.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.9.0) :tada:
